### ISSUE TITLE
v2 performance issues

### ISSUE THREAD

**Issue by ahill00** [ansible/ansible#13023]
##### Issue Type:
- Bug Report
##### Ansible Version:

$ ansible --version
ansible 2.0.0 (devel eb8ec73ff3) last updated 2015/11/03 14:30:56 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 22c2789b72) last updated 2015/10/28 23:03:02 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD c685b3d387) last updated 2015/10/28 23:03:10 (GMT -400)
  config file =
  configured module search path = None
##### Ansible Configuration:

none
##### Environment:

Mac OS X 10.10.4
##### Summary:

v1 playbook completes in ~3s.
v2 playbook doesn't complete.
##### Steps To Reproduce:

Playbook here: https://gist.github.com/andyhky/8c5ec9485eb059cf0226 but calling with a --limit which goes down to 100 hosts, e.g., --limit c0001-*compute[0:100] or [0-100] for v1. 

I am passing --forks 100 to ansible-playbook.

Total Inventory being passed has a little over 5K hosts across 15 groups, with the nova-compute group at ~400 with the --limit I'm passing.
##### Expected Results:

Ansible V1, completes in 3s.

v1 specifics:

$ ansible --version
ansible 1.8.2
  configured module search path = None
##### Actual Results:

It hangs, seemingly forever. I've let it run up to 58 minutes before giving up.  With ANSIBLE_DEBUG=true it hangs here:

 15102 1446579807.18443: got a task/handler to work on: TASK: debug msg={{ inventory_hostname }}
 15102 1446579807.18449: running TaskExecutor() for removed/TASK: debug msg={{ inventory_hostname }}
 15102 1446579807.18453: in run()
 15102 1446579807.18500: calling self._execute()
 15102 1446579807.18600: no remote address found for delegated host localhost
using its name, so success depends on DNS resolution
 15102 1446579807.18921: starting attempt loop
 15102 1446579807.18929: running the handler
 15102 1446579807.18935: handler run complete
 15102 1446579807.18943: attempt loop complete, returning result
 15102 1446579807.18948: _execute() done
 15102 1446579807.18952: dumping result to json
 15102 1446579807.18957: done dumping result, returning
 15102 1446579807.18964: done running TaskExecutor() for removed/TASK: debug msg=removed
 15102 1446579807.18966: sending task result
 15102 1446579807.18972: done sending task result

Which I think is somewhere around this part in the code:
https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/__init__.py#L296-L309

If I lower the --forks, the playbook ends up completing, but slowly.

**Comment by evanccnyc** [comment]
This could be related to https://github.com/ansible/ansible/issues/12239

**Comment by ahill00** [comment]
attempted to run from https://github.com/jimi-c/ansible/tree/issue_12477_performance_improvements_hostvars , still doesn't complete.

**Comment by Yannig** [comment]
@andyhky I have done a PR in order to fix a part of Ansible v2 performance. You can retrieve it here: https://github.com/ansible/ansible/pull/13000

Maybe you can give it a try.

**Comment by sivel** [comment]
@andyhky earlier told me that #13000 did not improve performance for him either.

```
13:16:37< ~andyhky> i'm hitting some perf issues with ansiblev2 and this playbook: https://gist.github.com/andyhky/8c5ec9485eb059cf0226 , if I run it with 100 forks against 100 hosts in v1 it finishes in 3s.
                    it doesn't complete in < 50m on v2.
13:18:11< sivel> andyhky: might be interested in both https://github.com/ansible/ansible/issues/12239 and https://github.com/ansible/ansible/pull/13000
13:18:33< ~andyhky> sivel: thanks!
13:18:56< sivel> andyhky: np :)
13:28:33< ~andyhky> sivel: fwiw, no noticable improvement w/ those patches
```

**Comment by Yannig** [comment]
@sivel @andyhky I will take a look if I can improve something.

**Comment by Yannig** [comment]
@sivel @andyhky This is not perfect but this patch may improve things.

**Comment by ahill00** [comment]
@sivel @Yannig - I tried c0ebb1c , still no improvement.

**Comment by mgedmin** [comment]
Could you (a) limit your test case to something reasonable (~5 minutes) and (b) run ansible 2 under a Python profiler, to see what it's doing?

Somebody ought to add a "How to profile Ansible" section somewhere, but I don't know where.  CONTRIBUTING.rst seems inappropriate somehow.  Anyway, here's [how I did it once](https://github.com/ansible/ansible/issues/11981#issuecomment-134578714):

```
$ pip install --user line-profiler runsnakerun
$ . ~/src/ansible/hacking/env-setup
$ python -m kernprof -o /tmp/ansible.prof ~/src/ansible/bin/ansible-playbook test.yml 
$ python -m pstats /tmp/ansible.prof
% sort cumtime
% stats 20
```

([RunSnakeRun](https://pypi.python.org/pypi/RunSnakeRun) is also sometimes useful for viewing Python profiles.)

**Comment by jimi-c** [comment]
@andyhky I believe the patch @bcoca applied yesterday should mostly resolve this. The underlying issue is that we're forking processes and either something is causing the memory to be modified (resulting in a copy-on-write of that RAM) or it is truly the minimal resident size at that point. For example, in my testing with a simple configuration, each forked process was taking up around 18MB of RAM. When I added a lot of variables to `group_vars/all`, the per-process RAM usage went up to the high 20MB's.

@bcoca's patch (0abeff45115689cd8e1262c8c666c9263977c499) re-implements the v1 behavior of forks being limited to the # of inventory hosts, however if you have 1000 hosts and run with `-f 1000` you'll still need more than 16GB of RAM (at a minimum) right now. We're looking to see if there's anything we can do to fix that, but right now I'd watch top and tune forks to your RAM size.

**Comment by ahill00** [comment]
@jimi-c - Thanks for the update.

FWIW, this is with 100 hosts/forks, not 1000.  

Adjusting forks helped the completion:  (with 0abeff4 applied)

10 forks: 1m42s
50 forks: 1m45s
80 forks: 1m47s

**Comment by jimi-c** [comment]
@andyhky excellent. And I thought @bcoca had applied that yesterday but that was to a non-devel branch. It is now applied to devel, so I believe we can close this for now if you're good to go?

**Comment by Yannig** [comment]
Hi @bcoca @andyhky @jimi-c,

Just for your information, when using this PR https://github.com/ansible/ansible/pull/13000 and the following playbook:

``` yaml
---
- name: Echo the hostname
  hosts: "compute"
  gather_facts: false
  connection: local

  tasks:
    - debug: msg="{{ inventory_hostname }}"
      delegate_to: localhost
```

Ansible v2 finish the job in about 1.5s with 100 hosts. Here's some numbers:

| host count | time |
| --- | --- |
| 100 | 1.5s |
| 200 | 3.368 |
| 400 | 8.992s |
| 800 | 31s |
| 1000 | 49.102s |
| 1600 | 2m10.720s |

I know the PR need more test and may break things but I think there might be some optimisation idea within it.

**Comment by ahill00** [comment]
@mgedmin - 

```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001  136.225  136.225 {execfile}
        1    0.004    0.004  136.224  136.224 /Users/USER/venv/ansiblev2/bin/ansible-playbook:21(<module>)
        1    0.000    0.000  135.938  135.938 /Users/USER/venv/ansiblev2/lib/ansible/cli/playbook.py:79(run)
        1    0.000    0.000  132.094  132.094 /Users/USER/venv/ansiblev2/lib/ansible/executor/playbook_executor.py:61(run)
        1    0.000    0.000  131.074  131.074 /Users/USER/venv/ansiblev2/lib/ansible/executor/task_queue_manager.py:172(run)
        1    0.005    0.005  130.613  130.613 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/linear.py:144(run)
      452   40.937    0.091  111.187    0.246 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:152(get_hosts)
   266/90    0.012    0.000  104.154    1.157 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:175(get_vars)
  266/178   10.308    0.039  104.101    0.585 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:339(_get_magic_variables)
182395193   57.194    0.000   69.260    0.000 /Users/USER/venv/ansiblev2/lib/ansible/inventory/host.py:38(__eq__)
       88    0.001    0.000   43.580    0.495 /Users/USER/venv/ansiblev2/lib/ansible/vars/hostvars.py:64(__getitem__)
       88    0.001    0.000   26.987    0.307 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:391(_get_delegated_vars)
       88    0.001    0.000   25.860    0.294 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:125(add_tqm_variables)
       88    0.002    0.000   12.981    0.148 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:122(get_failed_hosts)
       88    0.003    0.000   12.879    0.146 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:119(get_hosts_remaining)
184046205   12.327    0.000   12.383    0.000 {isinstance}
     6018    0.015    0.000    6.883    0.001 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:612(basedir)
    83970    5.485    0.000    5.485    0.000 {posix.stat}
    18954    0.028    0.000    3.652    0.000 /Users/USER/venv/lib/python2.7/genericpath.py:38(isdir)
     6019    0.008    0.000    3.596    0.001 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:603(is_directory)
```

**Comment by ahill00** [comment]
@Yannig - what count of forks are you using with #13000 ? I'm not seeing that dramatic of a performance increase testing that code.

**Comment by amenonsen** [comment]
Since #12823 was closed in favour of this issue, I thought I'd mention that the non-destructive tests have barely been improved by these changes. Last I checked, they took 21m to run. I tried again just now, and they took 20m. This is as opposed to 12/13m with an earlier version of devel.

**Comment by mgedmin** [comment]
Fun exponential runtime (thank you @Yannig  for the data)!
![ekrano nuotrauka is 2015-11-06 10-07-30](https://cloud.githubusercontent.com/assets/159967/10992418/4ad5c07c-846e-11e5-8b8e-8440242defde.png)

And thank you @andyhky for the profile.  Looks like `_get_magic_variables()` needs some attention.  And the number of `Host.__eq__` comparisons seems seriously excessive.  I'm guessing `isinstance` is the one called from `Host.__eq__`.

Seriously, a full 50% (69.260s out of 136.225s) of the time is spent inside `Host.__eq__`.  Can we make sure Hosts are singletons (i.e. no two Host objects exist with the same name in the same process) and then drop the custom `__hash__`/`__eq__` methods?

**Comment by Yannig** [comment]
@mgedmin when using a huge inventory (1000 hosts), Ansible v2 spent most of the time in templating vars which are static. Particularly, the vars ansible_play_hosts, play_hosts, groups, ungrouped, group_names (full list is available here: https://github.com/Yannig/ansible/commit/ccbdd6229a247e51cba15d8913687f03580a80ac)

If you take vars like play_hosts or ansible_play_hosts, with large inventory, they became huge and you spend more and more time in copying them.

Maybe a good optimization could be to assign them to a list of static vars which are known to be common to all hosts.

**Comment by jimi-c** [comment]
@Yannig yes, this is a problem with the HostVars - the entire set of variables is run through the Templar to make sure the hostvars are completely rendered. This really degrades performance when there are a large number of variables, and is something we've identified as a problem.

I also was profiling earlier in the week and found the calls to get_hosts() for the magic variable was impacting things, but it seemed like it was more of a problem with the large # of results returned with large inventories when templated, rather than the time to actually gather that data from inventory (which is negligible, as it's cached).

**Comment by jimi-c** [comment]
The hostvars templating theory can be tested with this patch:

<pre>
diff --git a/lib/ansible/vars/hostvars.py b/lib/ansible/vars/hostvars.py
index 9f83342..3ce2127 100644
--- a/lib/ansible/vars/hostvars.py
+++ b/lib/ansible/vars/hostvars.py
@@ -68,8 +68,9 @@ class HostVars(collections.Mapping):
 
         host = self._lookup.get(host_name)
         data = self._variable_manager.get_vars(loader=self._loader, host=host, play=self._play, include_hostvars=False)
-        templar = Templar(variables=data, loader=self._loader)
-        return templar.template(data, fail_on_undefined=False)
+        #templar = Templar(variables=data, loader=self._loader)
+        #return templar.template(data, fail_on_undefined=False)
+        return data
 
     def __contains__(self, host_name):
         item = self.get(host_name)
</pre>

It may be that, since we template things so late in the executor pipeline now, we no longer need to template within the hostvars themselves. This will take some testing to verify though.

**Comment by ahill00** [comment]
With the patch @jimi-c posted- I cancelled running after a couple of minutes:

```
         352701452 function calls (351358671 primitive calls) in 140.526 seconds

   Ordered by: cumulative time
   List reduced from 2141 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.002    0.002  140.526  140.526 {execfile}
        1    0.004    0.004  140.525  140.525 /Users/USER/venv/ansiblev2/bin/ansible-playbook:21(<module>)
        1    0.000    0.000  140.212  140.212 /Users/USER/venv/ansiblev2/lib/ansible/cli/playbook.py:79(run)
        1    0.000    0.000  136.676  136.676 /Users/USER/venv/ansiblev2/lib/ansible/executor/playbook_executor.py:61(run)
        1    0.001    0.001  135.620  135.620 /Users/USER/venv/ansiblev2/lib/ansible/executor/task_queue_manager.py:171(run)
        1    0.004    0.004  134.605  134.605 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/linear.py:144(run)
      427   42.764    0.100  116.201    0.272 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:159(get_hosts)
   251/85    0.011    0.000  108.023    1.271 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:175(get_vars)
  251/168    9.219    0.037  107.969    0.643 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:339(_get_magic_variables)
171963913   59.545    0.000   72.173    0.000 /Users/USER/venv/ansiblev2/lib/ansible/inventory/host.py:38(__eq__)
       83    0.002    0.000   46.797    0.564 /Users/USER/venv/ansiblev2/lib/ansible/vars/hostvars.py:64(__getitem__)
       83    0.002    0.000   28.324    0.341 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:391(_get_delegated_vars)
       83    0.001    0.000   26.590    0.320 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:129(add_tqm_variables)
       83    0.002    0.000   13.341    0.161 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:126(get_failed_hosts)
       83    0.004    0.000   13.248    0.160 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:123(get_hosts_remaining)
173533851   12.688    0.000   12.749    0.000 {isinstance}
     6003    0.017    0.000    7.135    0.001 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:619(basedir)
    83953    5.211    0.000    5.211    0.000 {posix.stat}
    75359    0.080    0.000    3.551    0.000 /Users/USER/venv/lib/python2.7/posixpath.py:350(abspath)
        1    0.000    0.000    3.178    3.178 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:51(__init__)
```

**Comment by jimi-c** [comment]
Interesting, only 427 calls to get_hosts() but each taking .1s so it adds up :/

**Comment by ahill00** [comment]
today's run:

```
ansible 2.0.0 (devel 7fd8b86a10) last updated 2015/11/11 15:51:17 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 22c2789b72) last updated 2015/10/28 23:03:02 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD c685b3d387) last updated 2015/10/28 23:03:10 (GMT -400)
  config file =
  configured module search path = Default w/o overrides
```

```
Wed Nov 11 16:24:30 2015    /tmp/ansible.prof

         7093843 function calls (7062438 primitive calls) in 1043.661 seconds

   Ordered by: cumulative time
   List reduced from 2144 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001 1043.662 1043.662 {execfile}
        1    0.010    0.010 1043.661 1043.661 /Users/user/venv/ansiblev2/bin/ansible-playbook:21(<module>)
        1    0.000    0.000 1043.058 1043.058 /Users/user/venv/ansiblev2/lib/ansible/cli/playbook.py:86(run)
        1    0.025    0.025 1039.634 1039.634 /Users/user/venv/ansiblev2/lib/ansible/executor/playbook_executor.py:66(run)
        1    0.001    0.001 1039.105 1039.105 /Users/user/venv/ansiblev2/lib/ansible/executor/task_queue_manager.py:167(run)
        1    0.004    0.004 1038.269 1038.269 /Users/user/venv/ansiblev2/lib/ansible/plugins/strategy/linear.py:144(run)
        2    1.679    0.839 1011.171  505.586 /Users/user/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:318(_wait_on_pending_results)
    88999 1007.320    0.011 1007.320    0.011 {time.sleep}
  305/103    0.012    0.000   24.851    0.241 /Users/user/venv/ansiblev2/lib/ansible/vars/__init__.py:174(get_vars)
  305/204   14.658    0.048   24.724    0.121 /Users/user/venv/ansiblev2/lib/ansible/vars/__init__.py:345(_get_magic_variables)
      101    0.559    0.006   14.550    0.144 /Users/user/venv/ansiblev2/lib/ansible/vars/hostvars.py:77(__getitem__)
     6057    0.016    0.000    6.872    0.001 /Users/user/venv/ansiblev2/lib/ansible/inventory/__init__.py:638(basedir)
    84070    5.293    0.000    5.293    0.000 {posix.stat}
    89201    0.499    0.000    4.204    0.000 /Users/user/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:185(_process_pending_results)
    89201    0.206    0.000    3.704    0.000 /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/queues.py:145(empty)
    89201    3.499    0.000    3.499    0.000 {method 'poll' of '_multiprocessing.Connection' objects}
    75412    0.074    0.000    3.369    0.000 /Users/user/venv/lib/python2.7/posixpath.py:350(abspath)
        1    0.000    0.000    3.060    3.060 /Users/user/venv/ansiblev2/lib/ansible/inventory/__init__.py:54(__init__)
        1    0.009    0.009    3.060    3.060 /Users/user/venv/ansiblev2/lib/ansible/inventory/__init__.py:90(parse_inventory)
    19009    0.029    0.000    3.017    0.000 /Users/user/venv/lib/python2.7/genericpath.py:38(isdir)
```

**Comment by mgedmin** [comment]
Wow, that's nice!  96.5% of the time is spent waiting for remote processes to do their work.

**Comment by bcoca** [comment]
that is actually a good place for the main process to be, network + running the tasks on remote should be the natural bottleneck.

**Comment by jimi-c** [comment]
I've been working on tuning this up, and have identified several of the main bottlenecks that are causing v2 to be on average 2x slower than v1. I hope to have something figured out this weekend or early next week.

**Comment by ahill00** [comment]
@jimi-c - Got a development branch or anything available to test with?

**Comment by jimi-c** [comment]
@andyhky all performance improvements have been folded into rc1, with the exception of one, which I"m still waiting on a PR from the user who debugged it (which impacts large #'s of include_vars).

So testing against devel or rc1 should get you pretty close to 1.x performance, if not let us know.

**Comment by ahill00** [comment]
Gave devel a try today. Still not completing the playbook... 

These are all local processes (example playbook delegates to localhost) and v1 completes in 3-4 seconds.

``````
$ ansible --version
ansible 2.1.0 (detached HEAD a19e083d33) last updated 2015/12/09 16:59:07 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 0b5555b62c) last updated 2015/12/09 16:58:49 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD cbed642009) last updated 2015/12/09 16:58:50 (GMT -400)
  config file =
  configured module search path = Default w/o overrides````
``````

Here's the profiling output after giving up after 9 minutes.

```
Wed Dec  9 17:10:04 2015    /tmp/ansible.prof

         22909008 function calls (22879181 primitive calls) in 554.257 seconds

   Ordered by: cumulative time
   List reduced from 2250 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001  554.258  554.258 {execfile}
        1    0.002    0.002  554.257  554.257 /Users/USER/venv/ansiblev2/bin/ansible-playbook:21(<module>)
        1    0.000    0.000  553.942  553.942 /Users/USER/venv/ansiblev2/lib/ansible/cli/playbook.py:86(run)
        1    0.000    0.000  549.745  549.745 /Users/USER/venv/ansiblev2/lib/ansible/executor/playbook_executor.py:65(run)
        1    0.001    0.001  548.655  548.655 /Users/USER/venv/ansiblev2/lib/ansible/executor/task_queue_manager.py:173(run)
        1    1.488    1.488  544.049  544.049 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/linear.py:144(run)
        2   37.444   18.722  521.649  260.825 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:333(_wait_on_pending_results)
  1805795  442.890    0.000  442.890    0.000 {time.sleep}
  1805793   13.961    0.000   39.385    0.000 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:179(_process_pending_results)
  1805793    6.229    0.000   25.424    0.000 /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/queues.py:145(empty)
  1805793   19.195    0.000   19.195    0.000 {method 'poll' of '_multiprocessing.Connection' objects}
  204/103    0.036    0.000   15.424    0.150 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:175(get_vars)
      204    8.201    0.040   14.256    0.070 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:354(_get_magic_variables)
      415    0.962    0.002    4.830    0.012 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:162(get_hosts)
      101    0.004    0.000    3.930    0.039 /Users/USER/venv/ansiblev2/lib/ansible/vars/__init__.py:397(_get_delegated_vars)
     5956    0.068    0.000    3.888    0.001 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:640(basedir)
        1    0.000    0.000    3.323    3.323 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:54(__init__)
        1    0.011    0.011    3.323    3.323 /Users/USER/venv/ansiblev2/lib/ansible/inventory/__init__.py:90(parse_inventory)
    84021    3.152    0.000    3.152    0.000 {posix.stat}
      101    0.037    0.000    2.901    0.029 /Users/USER/venv/ansiblev2/lib/ansible/plugins/strategy/__init__.py:135(_queue_task)
```

**Comment by Yannig** [comment]
Can you provide additional information about this problem:
- which playbook?
- which inventory?

**Comment by jimi-c** [comment]
Generated an inventory with this script (which results in an inventory of over 7500 hosts):

<pre>
rm -f inv
for j in {1..15}; do
  printf "[group%02d]\n" $j
  for i in {0..500}; do 
    printf "c00%02d%05d ansible_connection=local\n" $j $i; 
  done
done >> inv
</pre>

And running this slightly modified version of your playbook:

<pre>
---
- name: Echo the hostname
  hosts: group01
  gather_facts: false
  connection: ssh
  tasks:
    - debug: msg="{{ inventory_hostname }}"
      delegate_to: localhost
</pre>

When running `--forks=100`, I definitely see some slowness, due to the fact that higher numbers of forks can put a lot of memory pressure on your system. On my system (core i7 with 8GB ram), anything over `--forks=50` starts to degrade performance.

When running with lower numbers of forks though, I get a pretty consistent result:

<pre>
# time ansible-playbook -vv -i inv test.yml --limit=c0001*[0:100] --forks=50
Using /etc/ansible/ansible.cfg as config file
1 plays in test.yml
PLAY [Echo the hostname] *******************************************************
TASK [debug msg={{ inventory_hostname }}] **************************************
ok: [c000100008 -> localhost] => {
    "changed": false, 
    "msg": "c000100008"
}
...
c000100098                 : ok=1    changed=0    unreachable=0    failed=0   
c000100099                 : ok=1    changed=0    unreachable=0    failed=0   
c000100100                 : ok=1    changed=0    unreachable=0    failed=0   
real    0m6.827s
user    0m16.221s
sys 0m3.408s
</pre>

The variance here is about +/- 0.5s, but usually around 6.5s.

In a follow-up, I generated some variables into group_vars/all with this script:

<pre>
# for i in {1..5000}; do printf "var%05d: %05d\n" $i $i; done > group_vars/all
</pre>

And this definitely caused some (expected) slowness:

<pre>
# time ansible-playbook -vv -i inv test.yml --limit=c0001*[0:100] --forks=50
Using /etc/ansible/ansible.cfg as config file
1 plays in test.yml
PLAY [Echo the hostname] *******************************************************
TASK [debug msg={{ inventory_hostname }}] **************************************
ok: [c000100008 -> localhost] => {
    "changed": false, 
    "msg": "c000100008"
}
...
c000100099                 : ok=1    changed=0    unreachable=0    failed=0   
c000100100                 : ok=1    changed=0    unreachable=0    failed=0   
real    0m32.825s
user    2m6.158s
sys 0m25.499s
</pre>

So I'm curious how many variables you have for your inventory hosts here, as this is pretty much the worst case scenario (when there are a lot of variables in group_vars/all, because EVERY host gets a copy of those variables).

**Comment by ahill00** [comment]
@jimi-c 
There are a little over 5K hosts in group_vars/all, each with 20-40 vars- e.g., group_vars['all'][0]|length.

@Yannig 

https://gist.github.com/andyhky/8c5ec9485eb059cf0226 is the playbook. I'm unable to share the inventory. Are there specific characteristics you're interested in?

**Comment by jimi-c** [comment]
@andyhky ok, so my last example is what I'd expect to see on your system, not sure why you're running into more slowness. Do you still see the problem if you reduce the number of forks?

**Comment by ahill00** [comment]
@jimi-c - Using 50 forks the playbook completes in 15s.

**Comment by Yannig** [comment]
@jimi-c: maybe we'd better add a check against max fork

**Comment by ahill00** [comment]
Not sure if this is useful, but another characteristic of the inventory is that hosts are in multiple groups:

e.g., one of the 'c0001-compute' hosts is in the 'compute' group and the 'c0001' group

**Comment by jimi-c** [comment]
@Yannig it's difficult to detect, as it depends somewhat on the memory usage of the main thread pre-fork, which can be anywhere from 1-3Mb, and then the more workers you have the more the dictionary of variables is being copied out to each worker, which equates to more CPU and memory usage as well.

I'll add a note to the docs we're working on for the 2.0 release that really unless you're working on a system with 16GB+ of RAM, --forks=50 is probably as high as you want to go for now.

**Comment by ahill00** [comment]
For what it's worth I reran this with a current devel  ( f39ad1f13a3fdc126e99bf3da45da1945f17e9c3 ) and times are great.

```
real    0m6.901s
user    0m9.452s
sys 0m4.968s
```

```
$ ansible --version
ansible 2.2.0 (devel f39ad1f13a) last updated 2016/04/30 19:03:06 (GMT -400)
  lib/ansible/modules/core: (detached HEAD fd59dccdd7) last updated 2016/01/14 09:53:02 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 38dfe23336) last updated 2016/01/14 09:53:03 (GMT -400)
  config file =
  configured module search path = Default w/o overrides
```
